src/lib.rs

```diff
@@ -285,6 +297,77 @@ pub fn listen_fds() -> io::Result<impl Iterator<Item = RawFd>> {
     Ok(listen_fds)
 }
 
+/// Checks for file descriptors passed by the service manager for socket
+/// activation.
+///
+/// The function returns an iterator over file descriptors, starting from
+/// `SD_LISTEN_FDS_START`. The number of descriptors is obtained from the
+/// `LISTEN_FDS` environment variable.
+///
+/// If the `unset_env` parameter is set, the `LISTEN_PID`, `LISTEN_FDS` and
+/// `LISTEN_FDNAMES` environment variable will be unset before returning.
+/// Child processes will not see the fdnames passed to this process. This is
+/// usually not necessary, as a process should only use the `LISTEN_FDS`
+/// variable if it is the PID given in `LISTEN_PID`.
+///
+/// Before returning, the file descriptors are set as `O_CLOEXEC`.
+///
+/// See [`sd_listen_fds_with_names(3)`][sd_listen_fds_with_names] for details.
+///
+/// [sd_listen_fds_with_names]: https://www.freedesktop.org/software/systemd/man/sd_listen_fds.html
+///
+/// # Example
+///
+/// ```no_run
+/// let socket = sd_notify::listen_fds().map(|mut fds| fds.next().expect("missing fd"));
+/// ```
+pub fn listen_fds_with_names(
+    unset_env: bool,
+) -> io::Result<impl ExactSizeIterator<Item = (RawFd, String)>> {
+    let listen_fds = listen_fds_internal(unset_env)?;
+    let _guard = UnsetEnvGuard {
+        name: "LISTEN_FDNAMES",
+        unset_env,
+    };
+    zip_fds_with_names(listen_fds, env::var("LISTEN_FDNAMES").ok())
+}
+
+/// Internal helper that is independent of listen_fds function, for testing purposes.
+fn zip_fds_with_names(
+    listen_fds: impl ExactSizeIterator<Item = RawFd>,
+    listen_fdnames: Option<String>,
+) -> io::Result<impl ExactSizeIterator<Item = (RawFd, String)>> {
+    let listen_fdnames = if let Some(names) = listen_fdnames {
+        names.split(':').map(|x| x.to_owned()).collect::<Vec<_>>()
```

Comment:
I'm not super familiar with this functionality, but shouldn't we parse `foo::bar` as `["foo", "unknown", "bar"]` instead of `["foo", "", "bar"]`?

Reply:
Gentle nudge @cbranch, can you take a look at this? Thanks!

Reply:
I verified the behaviour of the libsystemd function and it indeed uses an empty string, though systemd should always provide a non-empty name. I added a comment with a code example for posterity.